### PR TITLE
Enhancement of Error messages - Feature request #3230

### PR DESCRIPTION
**Pull request by springcoil**
Better initial energy documentation.

This doesn't quite work properly because I'm looping over a loop, and not collecting all of the data from it. 

Tested it locally using. 

```import pymc3 as pm

with pm.Model():
    pm.HalfNormal('a', sd=1, transform=None)
    pm.sample()
```

I wanted to put this here just to mark my territory :) 

Does anyone have any suggestions on wording and what to do around the loop aspect.

### CODE REVIEW

**Comment by springcoil** [comment]
I've made some improvements, it now handles sets of RV.names and tuples of the -inf. Do I need a test for this error? I think things could be cleaned up :) 
Inspired by #3230

File: pymc3/step_methods/hmc/base_hmc.py [hunk 1]
```diff
@@ -110,11 +110,18 @@ def astep(self, q0):
...
+            model_points = ()
+            for RV in model.basic_RVs:
+                RV.name, RV.logp(model.test_point)
+                RV_set.add(RV.name)
+                model_points = model_points + (RV.logp(model.test_point).item(),)
+                
             self.potential.raise_ok(self._logp_dlogp_func._ordering.vmap)
-            raise ValueError('Bad initial energy: %s. The model '
-                             'might be misspecified.' % start.energy)
+            import ipdb; ipdb.set_trace()
```
Comment (twiecki) [inline pymc3/step_methods/hmc/base_hmc.py]: break-point committed.

Comment (springcoil) [reply pymc3/step_methods/hmc/base_hmc.py]: Ahh crap

**Comment by springcoil** [comment]
Some tests are failing. Fixing this and checking again locally.

**Comment by springcoil** [comment]
Ok, so the tests pass locally. I've also added a regression test for this error message. @aseyboldt @twiecki Travis should pass so let me know if you're happy with this.

File: pymc3/step_methods/hmc/base_hmc.py [hunk 2]
```diff
@@ -53,10 +53,10 @@ def __init__(self, vars=None, scaling=None, step_scale=0.25, is_cov=False,
             `energy`, and `random` methods.
         **theano_kwargs: passed to theano functions
         """
-        model = modelcontext(model)
+        self.model = modelcontext(model)
```
Comment (ColCarroll) [inline pymc3/step_methods/hmc/base_hmc.py]: letting the model leak into objects causes trouble for various reasons - can you move this to the line after you check for finite energy?

Specifically:

```
if not np.isfinite(start.energy):
    model = modelcontext(model)
    ...
```

Comment (springcoil) [reply pymc3/step_methods/hmc/base_hmc.py]: Is it ok to still have self.model = modelcontext(model)? Or do I need to remove that completely?

Comment (ColCarroll) [reply pymc3/step_methods/hmc/base_hmc.py]: I'd remove it unless there is a good reason (dependencies get really messy when every object always has a pointer to the model context)

Comment (springcoil) [reply pymc3/step_methods/hmc/base_hmc.py]: Otherwise I get an unbounded local variable.

Comment (ColCarroll) [reply pymc3/step_methods/hmc/base_hmc.py]: From what?

Comment (springcoil) [reply pymc3/step_methods/hmc/base_hmc.py]: If I do model = modelcontext(model) without declaring a self.model. I get an unbounded local variable

File: pymc3/step_methods/hmc/base_hmc.py [hunk 3]
```diff
@@ -110,11 +110,11 @@ def astep(self, q0):
...
         p0 = self.potential.random()
         start = self.integrator.compute_state(q0, p0)
-
+         
         if not np.isfinite(start.energy):
+            model = self._model
             self.potential.raise_ok(self._logp_dlogp_func._ordering.vmap)
-            raise ValueError('Bad initial energy: %s. The model '
-                             'might be misspecified.' % start.energy)
+            raise ValueError("Bad initial energy, check any log  probabilities that are inf or -inf: {}".format(model.check_test_point().to_string()))
```
Comment (ColCarroll) [inline pymc3/step_methods/hmc/base_hmc.py]:
```suggestion
            raise ValueError("Bad initial energy, check any log  probabilities that are inf or -inf:\n{}".format(model.check_test_point().to_string()))
```

**Review by ColCarroll — APPROVED** [review]
LGTM - suggested a small change in formatting. For what it is worth, output will now look something like 

```pythontb
Bad initial energy, check any log  probabilities that are inf or -inf:
mu_logodds__      -1.39
phi_log__         -0.77
likelihood     -4804.59
```

File: pymc3/step_methods/hmc/base_hmc.py (hunk 3 above)
Comment (springcoil) [reply pymc3/step_methods/hmc/base_hmc.py]: Thanks for the suggestion

File: pymc3/step_methods/hmc/base_hmc.py [hunk 4]
```diff
@@ -110,11 +110,11 @@ def astep(self, q0):
         """Perform a single HMC iteration."""
         p0 = self.potential.random()
         start = self.integrator.compute_state(q0, p0)
-
+
```
Comment (twiecki) [inline pymc3/step_methods/hmc/base_hmc.py]:
```suggestion

```

File: pymc3/step_methods/hmc/base_hmc.py [hunk 5]
```diff
@@ -110,11 +110,11 @@ def astep(self, q0):
...
         p0 = self.potential.random()
         start = self.integrator.compute_state(q0, p0)
-
+         
         if not np.isfinite(start.energy):
+            model = self._model
             self.potential.raise_ok(self._logp_dlogp_func._ordering.vmap)
-            raise ValueError('Bad initial energy: %s. The model '
-                             'might be misspecified.' % start.energy)
+            raise ValueError("Bad initial energy, check any log  probabilities that are inf or -inf:\n{}".format(model.check_test_point().to_string()))
```
Comment (twiecki) [inline pymc3/step_methods/hmc/base_hmc.py]:
```suggestion
            raise ValueError("Bad initial energy, check any log probabilities that are inf or -inf:\n{}".format(model.check_test_point().to_string()))
```

**Comment by aseyboldt** [comment]
I like this :-)
Maybe should print the list of bad variables though logging instead of putting them into the error message? Just as a matter of convention I usually like the error message itself to be short and only one line. We also show the convergence warnings though logging.
It is possible (I think) that no variable is inf or nan, and you still get a bad initial energy error. The problem could come from the kinetic energy (`x^TCx` for position `x` and covariance C` so that could overflow somewhere). Also, the problem might be with one of the transformations theano does on the logp graph, and that can change if you only compute part of the logp.

**Comment by springcoil** [comment]
Hi Adrian,
I like your suggestion but I think this is an informative error message and
printing it is the best course of action for people say using jupyter
notebooks. I know it's a long error message but it's an informative one.

**Comment by aseyboldt** [comment]
The logging output shows up in the notebook :-)

**Comment by springcoil** [comment]
Haha. Fair enough then. Do you want me to add it to logging instead?

**Comment by twiecki** [comment]
Let's add it to logging.

**Comment by springcoil** [comment]
Ok cool. I'm a bit swamped for the next 24 hours but will get to it then.

**Comment by aseyboldt** [comment]
Ah and another thing (sorry):
`model.check_test_point` gives us _all_ variables. If we have a 20_000 parameter model that isn't all that fun. Maybe we should filter out variables that are neither inf, nan or very large or very small (ie abs(logp) > 1e20 or so) and also limit it to maybe 100 items (so take `vals_series.head(100)`).

**Comment by springcoil** [comment]
Ahh yes. That's a valid point. I'm sure you can just filter on the data
frame!

**Comment by springcoil** [comment]
Hi guys, the error looks a bit poor to me in the logging. I'd love some help with this :) @aseyboldt What do you think?

**Comment by twiecki** [comment]
Screenshot?

**Comment by springcoil** [comment]
screenshot 2018-10-25 10 21 18

- It repeats over and over again.

File: pymc3/step_methods/hmc/base_hmc.py [hunk 6]
```diff
@@ -110,11 +110,12 @@ def astep(self, q0):
         """Perform a single HMC iteration."""
         p0 = self.potential.random()
         start = self.integrator.compute_state(q0, p0)
-
+        model = self._model
+        check_test_point = model.check_test_point()
+        error_logp = check_test_point.loc[(np.abs(check_test_point) >= 1e20) | ~(np.isfinite(check_test_point)) | np.isnan(check_test_point)]
```
Comment (aseyboldt) [inline pymc3/step_methods/hmc/base_hmc.py]: np.isfinite already checks if it is nan: `np.isfinite(np.nan) == False`

**Comment by aseyboldt** [comment]
This is in the terminal? If so, maybe we should change our default formatter :-)

**Comment by springcoil** [comment]
Yes @aseyboldt it's in the terminal. Is my logger wrong?

**Comment by springcoil** [comment]
My logger is `logger = logging.getLogger('pymc3')` should the pymc3 be __name__ ?

File: pymc3/step_methods/hmc/base_hmc.py (hunk 6 above)
Comment (springcoil) [reply pymc3/step_methods/hmc/base_hmc.py]: Ok I've removed that - thanks for that :)

**Comment by aseyboldt** [comment]
That logger should be fine. I'll play a bit with the extra argument to the logging function, give me a few minutes...

**Comment by springcoil** [comment]
Nw :)

**Comment by springcoil** [comment]
My concern was the fact that it showed the same warning over and over again. Surely just once is sufficient.

**Comment by aseyboldt** [comment]
I just realised that the test are going to fail because they still check the error message instead of the logging output. You can use this to get the logging stuff: https://docs.pytest.org/en/latest/logging.html#caplog-fixture

**Comment by springcoil** [comment]
Hi @aseyboldt thanks for catching that. I've updated the test to better logging. I had no idea about that fixture :)

**Pull request merged**